File: aoc-2023-rust/src/bin/01_02.rs
use std::io::{self, BufRead};

fn main() {
    let words = [
        "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];

    let mut ans = 0;

    for line in io::stdin().lock().lines() {
        let haystack = line.unwrap();

        let mut calibrations = vec![];

        // parse digits
        for (i, ch) in haystack.chars().enumerate() {
            if ch.is_digit(10) {
                calibrations.push((i, ch.to_digit(10).unwrap()));
            }
        }

        // find word matches
        for (nb, w) in words.iter().enumerate() {
            match haystack.find(w) {
                Some(i) => {
                    calibrations.push((i, nb as u32));
                }
                None => (),
            }
            match haystack.rfind(w) {
                Some(i) => {
                    calibrations.push((i, nb as u32));
                }
                None => (),
            }
        }

        calibrations.sort_by(|a, b| a.0.cmp(&b.0));

        // find first and last digits
        let n = calibrations.first().unwrap();
        let m = calibrations.last().unwrap();

        // build the number
        ans += n.1 * 10 + m.1;
    }

    println!("{}", ans);
}

File: typescript/src/main/2023/01_02.ts
import { kmpSearch } from "../utils/kmpSearch.ts";

export function solve(input: string): number {
  const space: [string, number][] = [
    ["0", 0],
    ["1", 1],
    ["2", 2],
    ["3", 3],
    ["4", 4],
    ["5", 5],
    ["6", 6],
    ["7", 7],
    ["8", 8],
    ["9", 9],
    ["zero", 0],
    ["one", 1],
    ["two", 2],
    ["three", 3],
    ["four", 4],
    ["five", 5],
    ["six", 6],
    ["seven", 7],
    ["eight", 8],
    ["nine", 9],
  ];

  let calibration = 0;
  const rows = input.trim().split("\n");
  for (const row of rows) {
    let ilhs = row.length - 1;
    let vlhs = NaN;

    let irhs = 0;
    let vrhs = NaN;

    for (const [needle, value] of space) {
      const matches = kmpSearch(row, needle);
      if (matches.length > 0) {
        const first = matches[0];
        if (first <= ilhs) {
          ilhs = first;
          vlhs = value;
        }

        const last = matches[matches.length - 1];
        if (last >= irhs) {
          irhs = last;
          vrhs = value;
        }
      }
    }

    calibration += vlhs * 10 + vrhs;
  }

  return calibration;
}
